client/config/config.go

```diff
@@ -0,0 +1,101 @@
+package config
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/cosmos/cosmos-sdk/client"
+)
+
+// Default constants
+const (
+	chainID        = ""
+	keyringBackend = "os"
+	output         = "text"
+	node           = "tcp://localhost:26657"
+	broadcastMode  = "sync"
+)
+
+type ClientConfig struct {
+	ChainID        string `mapstructure:"chain-id" json:"chain-id"`
+	KeyringBackend string `mapstructure:"keyring-backend" json:"keyring-backend"`
+	Output         string `mapstructure:"output" json:"output"`
+	Node           string `mapstructure:"node" json:"node"`
+	BroadcastMode  string `mapstructure:"broadcast-mode" json:"broadcast-mode"`
+}
+
+// DefaultClientConfig returns the reference to ClientConfig with default values.
+func DefaultClientConfig() *ClientConfig {
+	return &ClientConfig{chainID, keyringBackend, output, node, broadcastMode}
+}
+
+func (c *ClientConfig) SetChainID(chainID string) {
+	c.ChainID = chainID
+}
+
+func (c *ClientConfig) SetKeyringBackend(keyringBackend string) {
+	c.KeyringBackend = keyringBackend
+}
+
+func (c *ClientConfig) SetOutput(output string) {
+	c.Output = output
+}
+
+func (c *ClientConfig) SetNode(node string) {
+	c.Node = node
+}
+
+func (c *ClientConfig) SetBroadcastMode(broadcastMode string) {
+	c.BroadcastMode = broadcastMode
+}
+
+// ReadFromClientConfig reads values from client.toml file and updates them in client Context
+func ReadFromClientConfig(ctx client.Context) (client.Context, error) {
+	configPath := filepath.Join(ctx.HomeDir, "config")
+	configFilePath := filepath.Join(configPath, "client.toml")
+	conf := DefaultClientConfig()
+
+	// if config.toml file does not exist we create it and write default ClientConfig values into it.
+	if _, err := os.Stat(configFilePath); os.IsNotExist(err) {
+		if err := ensureConfigPath(configPath); err != nil {
+			return ctx, fmt.Errorf("couldn't make client config: %v", err)
+		}
+
+		configTemplate, err := initConfigTemplate()
+		if err != nil {
+			return ctx, fmt.Errorf("could not initiate config template: %v", err)
+		}
+
+		if err := writeConfigFile(configFilePath, conf, configTemplate); err != nil {
+			return ctx, fmt.Errorf("could not write client config to the file: %v", err)
+		}
+	}
+
+	conf, err := getClientConfig(configPath, ctx.Viper)
+	if err != nil {
+		return ctx, fmt.Errorf("couldn't get client config: %v", err)
+	}
+	// we need to update KeyringDir field on Client Context first cause it is used in NewKeyringFromFlags
+	ctx = ctx.WithOutputFormat(conf.Output).
+		WithKeyringDir(ctx.HomeDir).
```

Comment:
I just noticed `client.Context` has both `KeyringDir` and `HomeDir`. Should we maybe add `conf.KeyringDir` too?

Reply:
Is `KeyringDir` ever passed through as a flag?

Reply:
> I just noticed `client.Context` has both `KeyringDir` and `HomeDir`. Should we maybe add `conf.KeyringDir` too?

Yeah I will go ahead and add new field KeyringDir to ClientConfig struct. It equals clientCtx by default, but it might be changed using config Command.

Reply:
> Is `KeyringDir` ever passed through as a flag?

flags.FlagKeyringDir this is persistant flag. But I never seen it is passed as a flag in reality.

Reply:
If I am not mistaken, the keyring directory option is used only by a subset of backends. It's more of an exception that an actual configuration option. I would suggest avoiding adding this config option to the client configuration.

Reply:
OK, I wasn't aware of this. Let's remove keyringDir then if it adds confusion.

Reply:
Done